### PR TITLE
Sys 4191 register chain handler

### PR DESCRIPTION
**Pull request by ivan-cholakov**
Initial implementation of `register_chain_handler` and `update_chain_handler`

### CODE REVIEW

File: pallets/avn-anchor/src/lib.rs [hunk 1]
```diff
@@ -6,62 +6,110 @@ extern crate alloc;
...
+    pub type ChainHandlers<T: Config> = StorageMap<_, Blake2_128Concat, ChainId, T::AccountId>;
 
     #[pallet::call]
     impl<T: Config> Pallet<T> {
         #[pallet::weight(<T as pallet::Config>::WeightInfo::register_chain_handler())]
         #[pallet::call_index(0)]
         pub fn register_chain_handler(
             origin: OriginFor<T>,
-            // handler:
+            chain_id: ChainId,
```
Comment (thadouk) [inline pallets/avn-anchor/src/lib.rs:76]: I would suggest the input here is a friendly name, and ChainId is a nonce managed by the pallet, ensuring uniqueness 
We store the chain friendly name with the pallet data.

File: pallets/avn-anchor/src/lib.rs [hunk 2]
```diff
@@ -6,62 +6,110 @@ extern crate alloc;
...
         #[pallet::call_index(1)]
         pub fn update_chain_handler(
             origin: OriginFor<T>,
-            // handler:
+            chain_id: ChainId,
+            new_handler_account_id: T::AccountId,
         ) -> DispatchResult {
+            ensure_signed(origin)?;
+
+            ChainHandlers::<T>::try_mutate(&chain_id, |maybe_handler| -> DispatchResult {
```
Comment (thadouk) [inline pallets/avn-anchor/src/lib.rs:102]: We need to add a check here to ensure that the sender of this is actually the handler account otherwise return an error.

File: pallets/avn-anchor/src/lib.rs [hunk 3]
```diff
@@ -6,62 +6,110 @@ extern crate alloc;
...
+                let handler = maybe_handler.as_mut().ok_or(Error::<T>::HandlerNotRegistered)?;
```
Comment (thadouk) [inline pallets/avn-anchor/src/lib.rs:103]: I think if the data structure is chain_id -> handler if the chain id is not returning a handler then we should error with something that says that "chain_id is not registered" rather that the handler was not found.

File: pallets/avn-anchor/src/tests.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            AvnAnchor::update_chain_handler(
+                RuntimeOrigin::signed(new_handler),
+                chain_id,
+                new_handler
+            ),
+            Error::<TestRuntime>::HandlerNotRegistered
+        );
+    });
+}
+
```
Comment (thadouk) [inline pallets/avn-anchor/src/tests.rs:118]: we also need a test here that tests that a handler that attempts an update is possible only if the sender is the existing handler

File: pallets/avn-anchor/src/lib.rs [hunk 5]
```diff
@@ -6,62 +6,129 @@ extern crate alloc;
 use alloc::string::String;
 
 // use pallet_avn::{self as avn};
```
Comment (nahuseyoum) [inline pallets/avn-anchor/src/lib.rs:8]:
```suggestion
```

File: pallets/avn-anchor/src/lib.rs [hunk 6]
```diff
@@ -6,62 +6,129 @@ extern crate alloc;
...
         #[pallet::weight(<T as pallet::Config>::WeightInfo::register_chain_handler())]
         #[pallet::call_index(0)]
         pub fn register_chain_handler(
             origin: OriginFor<T>,
-            // handler:
+            name: BoundedVec<u8, ChainNameLimit>,
         ) -> DispatchResult {
+            let handler = ensure_signed(origin)?;
+
+            ensure!(!ChainHandlers::<T>::contains_key(&handler), Error::<T>::HandlerAlreadyRegistered);
```
Comment (nahuseyoum) [inline pallets/avn-anchor/src/lib.rs:94]: Can we validate that `name` is not null?

File: pallets/avn-anchor/src/lib.rs [hunk 7]
```diff
@@ -6,62 +6,129 @@ extern crate alloc;
...
-            // handler:
+            new_handler: T::AccountId,
         ) -> DispatchResult {
+            let old_handler = ensure_signed(origin)?;
+
+            let chain_data = ChainHandlers::<T>::get(&old_handler).ok_or(Error::<T>::ChainNotRegistered)?;
+
+            ensure!(!ChainHandlers::<T>::contains_key(&new_handler), Error::<T>::HandlerAlreadyRegistered);
+
+            ChainHandlers::<T>::remove(&old_handler);
```
Comment (nahuseyoum) [inline pallets/avn-anchor/src/lib.rs:122]: Can't we use `mutate`?

File: pallets/avn-anchor/src/tests.rs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        assert_eq!(chain_data1.chain_id, 0);
+        assert_eq!(chain_data1.name, name1);
+        assert_eq!(chain_data2.chain_id, 1);
+        assert_eq!(chain_data2.name, name2);
+
+        System::assert_has_event(Event::ChainHandlerRegistered(handler1, 0, name1).into());
+        System::assert_has_event(Event::ChainHandlerRegistered(handler2, 1, name2).into());
+    });
+}
```
Comment (nahuseyoum) [inline pallets/avn-anchor/src/tests.rs:143]: Can we add a test for null chain name?

**Review by nahuseyoum — APPROVED** [review]

File: pallets/avn-anchor/src/lib.rs [hunk 9]
```diff
@@ -6,62 +6,129 @@ extern crate alloc;
...
             Ok(())
         }
 
-        #[pallet::weight( <T as pallet::Config>::WeightInfo::update_chain_handler())]
+        #[pallet::weight(<T as pallet::Config>::WeightInfo::update_chain_handler())]
         #[pallet::call_index(1)]
         pub fn update_chain_handler(
             origin: OriginFor<T>,
-            // handler:
+            new_handler: T::AccountId,
```
Comment (thadouk) [inline pallets/avn-anchor/src/lib.rs:114]: Do you think that we should have a proof attached here with signed data of the new_handler?

**Review by thadouk — APPROVED** [review]

File: pallets/avn-anchor/src/lib.rs (hunk 9 above)
Comment (ivan-cholakov) [reply pallets/avn-anchor/src/lib.rs:113]: Not sure this is needed since the only actor that can set the new handler is the current handler
